### PR TITLE
Code optimizations / refactor

### PR DESCRIPTION
**Pull request by GermanAizek**
Hello everyone, this PR is basically refactoring code and minor compiler-level optimizations. I expect feedback.
@sfan5 @SmallJoker @rubenwardy @Wuzzy2 

## How to test
Building MSVC 17 2022 and run configuration Debug on Win10 x64

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -533,7 +533,7 @@ void ServerMap::transformLiquids(std::map<v3s16, MapBlock*> &modified_blocks,
 		infostream<<"transformLiquids(): initial_size="<<initial_size<<std::endl;*/
 
 	// list of nodes that due to viscosity have not reached their max level height
-	std::deque<v3s16> must_reflow;
+	std::vector<v3s16> must_reflow;
```
Comment (SmallJoker) [inline src/map.cpp:536]: Why? Deque is more efficient for insertions.

File: src/network/connection.cpp [hunk 2]
```diff
@@ -64,7 +64,7 @@ BufferedPacketPtr makePacket(Address &address, const SharedBuffer<u8> &data,
 {
 	u32 packet_size = data.getSize() + BASE_HEADER_SIZE;
 
-	BufferedPacketPtr p(new BufferedPacket(packet_size));
+	auto p = std::make_shared<BufferedPacket>(packet_size);
```
Comment (SmallJoker) [inline src/network/connection.cpp:67]: `BufferedPacketPtr` makes it more readable. No need to change this.

File: src/network/connection.cpp [hunk 3]
```diff
@@ -492,10 +492,10 @@ SharedBuffer<u8> IncomingSplitBuffer::insert(BufferedPacketPtr &p_ptr, bool reli
 
 void IncomingSplitBuffer::removeUnreliableTimedOuts(float dtime, float timeout)
 {
-	std::deque<u16> remove_queue;
+	std::vector<u16> remove_queue;
```
Comment (SmallJoker) [inline src/network/connection.cpp:495]: Deuque is more efficient for insertions (push_back).

File: src/script/common/c_content.cpp [hunk 4]
```diff
@@ -1447,9 +1447,9 @@ ToolCapabilities read_tool_capabilities(
 		int table_groupcaps = lua_gettop(L);
 		lua_pushnil(L);
 		while(lua_next(L, table_groupcaps) != 0){
-			// key at index -2 and value at index -1
-			std::string groupname = luaL_checkstring(L, -2);
 			if(lua_istable(L, -1)){
+				// key at index -2 and value at index -1
+				std::string groupname = luaL_checkstring(L, -2);
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp:1452]: This contains a check. Previous version was better.

File: src/script/cpp_api/s_base.cpp [hunk 5]
```diff
@@ -159,19 +159,25 @@ int ScriptApiBase::luaPanic(lua_State *L)
...
-	static const std::vector<std::pair<std::string, lua_CFunction>> m_libs = {
-		{ "", luaopen_base },
-		{ LUA_TABLIBNAME,  luaopen_table   },
-		{ LUA_OSLIBNAME,   luaopen_os      },
-		{ LUA_STRLIBNAME,  luaopen_string  },
-		{ LUA_MATHLIBNAME, luaopen_math    },
-		{ LUA_DBLIBNAME,   luaopen_debug   },
 #if USE_LUAJIT
-		{ LUA_JITLIBNAME,  luaopen_jit     },
+#define M_LIBS_SIZE 7
```
Comment (SmallJoker) [inline src/script/cpp_api/s_base.cpp:163]: This optimization is somewhat useless. It's called once per session.

File: src/tool.cpp [hunk 6]
```diff
@@ -33,9 +33,9 @@ void ToolGroupCap::toJson(Json::Value &object) const
 	object["uses"] = uses;
 
 	Json::Value times_object;
-	for (auto time : times)
+	for (const auto &time : times)
```
Comment (SmallJoker) [inline src/tool.cpp:36]: Useless reference to iterator.

File: src/client/clientmap.cpp [hunk 7]
```diff
@@ -840,9 +840,7 @@ void ClientMap::updateDrawListShadow(v3f shadow_light_pos, v3f shadow_light_dir,
 	v3s16 p_blocks_max;
 	getBlocksInViewRange(cam_pos_nodes, &p_blocks_min, &p_blocks_max, radius + length);
 
-	std::vector<v2s16> blocks_in_range;
-
-	for (auto &i : m_drawlist_shadow) {
+	for (const auto &i : m_drawlist_shadow) {
```
Comment (SmallJoker) [inline src/client/clientmap.cpp:843]: Useless reference to iterator.

File: src/client/fontengine.cpp [hunk 8]
```diff
@@ -90,7 +90,7 @@ void FontEngine::cleanCache()
 }
 
 /******************************************************************************/
-irr::gui::IGUIFont *FontEngine::getFont(FontSpec spec)
+irr::gui::IGUIFont *FontEngine::getFont(const FontSpec &spec)
```
Comment (SmallJoker) [inline src/client/fontengine.cpp:93]: `FontSpec` is only a few bytes big. Taking a reference to it is actually a worsening.

**Review by SmallJoker — COMMENTED** [review]
The improvements are negligible in terms of performance and not really needed. Nonetheless, it's an acceptable first-time PR - let's try to make the best out of it.

**Comment by GermanAizek** [comment]
@SmallJoker, thanks for code review, I will correct all your comments today

File: src/content/mod_configuration.cpp [hunk 9]
```diff
@@ -216,7 +216,7 @@ void ModConfiguration::resolveDependencies()
 
 	// Step 2: get dependencies (including optional dependencies)
 	// of each mod, split mods into satisfied and unsatisfied
-	std::list<ModSpec> satisfied;
+	std::vector<ModSpec> satisfied;
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/content/mod_configuration.cpp:219]: @SmallJoker,

- `satisfied` as `std::vector`
if you use 'std::vector', then algorithmic complexity of operations will not change, but sequential arrangement of elements in memory can improve performance.

- `unsatisfied` as `std::queue`
container implements FIFO queue, you can use 'std:: queue'.

File: src/database/database-dummy.cpp [hunk 10]
```diff
@@ -36,7 +36,7 @@ void Database_Dummy::loadBlock(const v3s16 &pos, std::string *block)
 	s64 i = getBlockAsInteger(pos);
 	auto it = m_database.find(i);
 	if (it == m_database.end()) {
-		*block = "";
+		(*block).clear();
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/database/database-dummy.cpp:39]: @SmallJoker,
yes, you noticed correctly, the compiler was silent without errors

File: src/map.cpp (hunk 1 above)
Comment (GermanAizek) [reply src/map.cpp:536]: @SmallJoker,

- `must_reflow` as `std::vector`

https://stackoverflow.com/questions/35483716/is-stddeque-faster-than-stdvector-for-inserting-at-the-end#answer-42743181

File: src/network/connection.cpp (hunk 2 above)
Comment (GermanAizek) [reply src/network/connection.cpp:67]: @SmallJoker,
https://github.com/ros-planning/moveit/issues/1411
https://stackoverflow.com/questions/20895648/difference-in-make-shared-and-normal-shared-ptr-in-c

File: src/network/connection.cpp (hunk 3 above)
Comment (GermanAizek) [reply src/network/connection.cpp:495]: @SmallJoker,

- `remove_queue` as `std::vector`

https://stackoverflow.com/questions/35483716/is-stddeque-faster-than-stdvector-for-inserting-at-the-end#answer-42743181

File: src/client/game.cpp [hunk 11]
```diff
@@ -91,7 +91,7 @@ struct TextDestNodeMetadata : public TextDest
 	// This is deprecated I guess? -celeron55
 	void gotText(const std::wstring &text)
 	{
-		std::string ntext = wide_to_utf8(text);
+		const std::string& ntext = wide_to_utf8(text);
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/client/game.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/database/database-dummy.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/database/database-dummy.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/filesys.cpp [hunk 12]
```diff
@@ -639,7 +639,7 @@ std::string RemoveLastPathComponent(const std::string &path,
 		std::string *removed, int count)
 {
 	if(removed)
-		*removed = "";
+		(*removed).clear();
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/filesys.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/gui/guiFormSpecMenu.cpp [hunk 13]
```diff
@@ -647,7 +647,6 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
 	std::vector<std::string> v_pos = split(parts[0],',');
 	std::vector<std::string> v_geom = split(parts[1],',');
 	std::string name = parts[3];
-	std::string value = parts[4];
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply src/gui/guiFormSpecMenu.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/script/common/c_content.cpp (hunk 4 above)
Comment (GermanAizek) [reply src/script/common/c_content.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/script/cpp_api/s_base.cpp (hunk 5 above)
Comment (GermanAizek) [reply src/script/cpp_api/s_base.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/tool.cpp (hunk 6 above)
Comment (GermanAizek) [reply src/tool.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/client/clientmap.cpp (hunk 7 above)
Comment (GermanAizek) [reply src/client/clientmap.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

File: src/client/fontengine.cpp (hunk 8 above)
Comment (GermanAizek) [reply src/client/fontengine.cpp]: @SmallJoker, fix it https://github.com/minetest/minetest/pull/12704/commits/7b18f1da0cf9bc70b334d51fc03ca93000e70811

**Comment by GermanAizek** [comment]
@SmallJoker,
Waiting for your response on change container type commits

**Comment by Zughy** [comment]
@GermanAizek there's no need to tag us in every comment, one tag after the push will suffice (also, we receive a notification anyway when someone replies to a review)

**Comment by GermanAizek** [comment]
@Zughy, ok sorry for spam

File: src/network/connection.cpp [hunk 14]
```diff
@@ -505,7 +505,7 @@ void IncomingSplitBuffer::removeUnreliableTimedOuts(float dtime, float timeout)
 				remove_queue.push_back(i.first);
 		}
 	}
-	for (u16 j : remove_queue) {
+	for (const auto &j : remove_queue) {
```
Comment (sfan5) [inline src/network/connection.cpp:508]: Taking a reference of a tiny variables will be less efficient than a copy

File: src/script/cpp_api/s_base.cpp [hunk 15]
```diff
@@ -160,18 +160,18 @@ int ScriptApiBase::luaPanic(lua_State *L)
...
-		{ LUA_DBLIBNAME,   luaopen_debug   },
+		{"", luaopen_base},
+		{LUA_TABLIBNAME, luaopen_table},
+		{LUA_OSLIBNAME, luaopen_os},
+		{LUA_STRLIBNAME, luaopen_string},
+		{LUA_MATHLIBNAME, luaopen_math},
+		{LUA_DBLIBNAME, luaopen_debug},
 #if USE_LUAJIT
-		{ LUA_JITLIBNAME,  luaopen_jit     },
+		{LUA_JITLIBNAME, luaopen_jit},
```
Comment (sfan5) [inline src/script/cpp_api/s_base.cpp:170]: the entries were all aligned before

File: src/content/mod_configuration.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/content/mod_configuration.cpp:219]: Either you're re-allocating an array of `ModSpec` in `std::vector` on append, or you have a doubly-linked list where front and back are each O(1). There's however no use of direct indexing here (only push_back & pop_back), so I don't see any improvement here.

File: src/network/connection.cpp (hunk 14 above)
Comment (GermanAizek) [reply src/network/connection.cpp:508]: @sfan5, yes, that's right, type is too small, inattentively changed to `const &`

File: src/script/cpp_api/s_base.cpp (hunk 15 above)
Comment (GermanAizek) [reply src/script/cpp_api/s_base.cpp:170]: clang-format file format settings were redefined and did not allow indentation

Comment (GermanAizek) [reply src/script/cpp_api/s_base.cpp]: fix it https://github.com/minetest/minetest/pull/12704/commits/50efdb084bf3c4a1af5757106c26a1ccbf34134d

File: src/network/connection.cpp (hunk 14 above)
Comment (GermanAizek) [reply src/network/connection.cpp]: fix it https://github.com/minetest/minetest/pull/12704/commits/50efdb084bf3c4a1af5757106c26a1ccbf34134d

**Comment by GermanAizek** [comment]
What could I have forgotten to change? I will not add new changes to this PR because it is large

**Review by SmallJoker — APPROVED** [review]
LGTM

**Review by sfan5 — APPROVED** [review]
LGTM

**Pull request merged**